File: 2389.和有限的最长子序列.py
#
# @lc app=leetcode.cn id=2389 lang=python3
#
# [2389] 和有限的最长子序列
#

# @lc code=start
class Solution:
    #from bisect import bisect_right
    def answerQueries(self, nums: List[int], queries: List[int]) -> List[int]:
        '''先把nums排序 再遍历一遍记录当前位置的总和
            二分查找大于等于queries[i]的第一个数'''
        nums.sort()
        for i in range(1,len(nums)):
            nums[i]=nums[i-1]+nums[i]
        
        #二分查找
        ans=[]
        for target in queries:
            idx=bisect_right(nums,target)
            ans.append(idx)
        return ans            
# @lc code=end

File: 2389.和有限的最长子序列.cpp
/*
 * @lc app=leetcode.cn id=2389 lang=cpp
 *
 * [2389] 和有限的最长子序列
 */

// @lc code=start
class Solution {
public:
    vector<int> answerQueries(vector<int>& nums, vector<int>& queries) {
        sort(nums.begin(), nums.end());
        for (int i = 1; i < nums.size();++i) {
            nums[i] = nums[i - 1] + nums[i];
        }

            vector<int> ans;
        for (int target : queries) {
            auto idx = upper_bound(nums.begin(), nums.end(), target);
            ans.push_back(idx-nums.begin());
        }
        return ans;
    }
};
// @lc code=end

